### ISSUE TITLE
Wrong filename for files generated by g++

### ISSUE THREAD

**Issue by royqh1979** [msys2/MINGW-packages#13797]
### Description

I run mingw-w64 gcc under cmd and GBK as the console's default charset.

I upgraded gcc from 11.2 to 12.2 recently.

When i use g++ to compiler a file that have non-ascii charater in it，for example:
`
g++.exe  "R:/测试.cpp" -o "R:/测试.exe" -O2 -pipe -s -static
`

The file generated by g++ is '²âÊÔ.exe', which should be '测试.exe'.

And i've tested gcc 12.2 compiled by winlibs, which doesn't have this problem.

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

Microsoft Windows [Ver 10.0.19043.2130]

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Expected behavior

When i use g++ to compiler a file that have non-ascii charater in it，for example:
`
g++.exe  "R:/测试.cpp" -o "R:/测试.exe" -O2 -pipe -s -static
`

The file generated by g++ should be '测试.exe'.

### Actual behavior

The file generated by g++ is actually '²âÊÔ.exe'

### Repro steps

1. open a cmd windows and set it's default charset to GBK.
2. use g++ to compiler a file that has non-ascii charaters in its name
3. The file generated by g++ has wrong file name.

**Comment by lb90** [comment]
There's has been recent rework of the file handling in binutils for Windows, see https://github.com/msys2/MINGW-packages/issues/12699. Moreover, I am also getting strange errors from gcc-ar when running Meson, such as:

[207/425] Linking static target gtk/libgtk.a
FAILED: gtk/libgtk.a
"gcc-ar" csrDT gtk/libgtk.a @gtk/libgtk.a.rsp
D:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ar.exe: gtk/libgtk.a: No such file or directory
[216/425] Compiling C object demos/gtk-demo/gtk4-demo.exe.p/combobox.c.obj
ninja: build stopped: subcommand failed.

Though both files exist. In that case I have to delete `gtk/libgtk.a` and `gtk/libgtk.a.rsp` manually and run ninja again.

I'll take a look at all those issues, but cannot guarantee anything

**Comment by lb90** [comment]
It could even be that gcc is using `argc` / `argv` directly, event though those shouldn't really be used on Windows

**Comment by revelator** [comment]
yeah places like this is where the toolchain bites us in the backside being a port from the posix world to windows there will be certain quirks like this. recently there was a guy who asked some questions about using the gnu toolchain for developing drivers on windows and running into a lot of snags because the toolchain did not support the correct options for windows driver development. atleast we now have ASLR and DEP :) only took about 20 years to get that ball running hehe.
